odbc-api/src/handles/columns.rs

```diff
@@ -0,0 +1,42 @@
+use super::{as_handle::AsHandle, buffer::buf_ptr, error::IntoResult, Statement, StatementImpl};
+use crate::{CursorImpl, Error};
+use odbc_sys::{SQLColumnsW, Stmt};
+use std::{borrow::BorrowMut, convert::TryInto};
+use widestring::U16Str;
+
+/// Shared implementation for executing a columns query between [`Connection`] and
+/// [`Preallocated`].
+pub fn columns<'o, S>(
```

Comment:
I would be more comfortable with a mutating method on `StatementImpl` which does not model the state transition to a cursor in the type system, yet (In my mental model of the code I seperated the layers, by what kind of decisions I would make in each of them). I would prefer introducing that in the above layer on `Connection` and `PreallocatedStatement`. The resulting top level implementation is probably small enough to be repeated, but could IMHO also be put in `execute.rs`. The whole situations to me seems analog to executing a special query with predefined parameters.

Reply:
Ok makes sense -- I think I'm starting to understand the structure better.

I updated the commit, I hope this is the right direction :)

Reply:
> I updated the commit, I hope this is the right direction :)

I hope so, too. Mostly because I think I am going to merge this later.